### PR TITLE
[RFC][FSDP2] Renamed `FSDP` to `FSDPModule`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124955
* #124787
* #124780
* #124768
* #124767
* #124741
* #124651

This PR renames the `FSDP` class to `FSDPModule`. This is a BC breaking change. The rationale is that `FSDPModule` is more descriptive since `fully_shard` is a module-level API (applied to a `module` arg), so the `FSDP` class will always correspond to a module.

Also, users commonly import `FullyShardedDataParallel` as `FSDP`, so this can help avoid some name conflict in some cases.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/distributed/_composable/fsdp/fully_shard.py [hunk 1]
```diff
@@ -136,7 +136,7 @@ def fully_shard(
     # Place FSDP leftmost for highest priority in the method resolution order
     cls = module.__class__
     dct = {"__deepcopy__": unimplemented_deepcopy}
-    new_cls = type(f"FSDP{cls.__name__}", (FSDP, cls), dct)
+    new_cls = type(f"FSDP{cls.__name__}", (FSDPModule, cls), dct)
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/fully_shard.py:139]: This won't affect the `__repr__` of the wrapped class iiuc?

**Review by wanchaol — APPROVED** [review]

**Comment by awgu** [comment]
cc: @wconstab If we land this change, then I think we need to change the PiPPy code to migrate `FSDP` -> `FSDPModule`.

**Review by wconstab — APPROVED** [review]
lgtm, thanks for the heads up RE pippy, i dont mind changing that to match.

**Comment by weifengpy** [comment]
oops. now I will use try-except in TorchTune to import FSDPModule

**Pull request closed without merging**